### DISCUSSION TITLE
Help needed: VsCode debug config

### DISCUSSION THREAD

**Discussion by philwilson** [cargo-lambda/cargo-lambda#426] [Q&A]
Hi, I would like to ask about how to get the debugging to work in vscode when using cargo-lambda.

Usually I would use the default launch.json template created which would look something like this.

```
{
      "type": "lldb",
      "request": "launch",
      "name": "Debug executable 'lambda-rust-hello-world'",
      "cargo": {
        "args": [
          "build",
          "--bin=lambda-rust-hello-world",
          "--package=lambda-rust-hello-world"
        ],
        "filter": {
          "name": "lambda-rust-hello-world",
          "kind": "bin"
        }
      },
      "args": [],
      "cwd": "${workspaceFolder}"
    },
```

it is using the codeLLDB extension to manage debugging.

I have gone through all of the iterations of configuration that I can think of to get the debug to work with the ``cargo lambda watch`` command

Any guidance would be appreciated.

**Comment by calavera** [comment]
Did you read this already? https://www.cargo-lambda.info/commands/watch.html#debug-with-breakpoints

**Comment by ngryman** [comment]
Hi,

Thanks for this awesome library! Unfortunately, I don't manage to make debugging work either...
I've read the above documentation, but for some reason, the process gets stuck at startup.

I'm starting the `cargo lambda` aside as follows:

```sh
$ cargo lambda watch --only-lambda-apis
```

And here is my `launch.json`:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug 'analyzer' function",
      "cargo": {
        "args": ["run", "--package=analyzer"],
        "env": {
          "AWS_LAMBDA_FUNCTION_VERSION": "1",
          "AWS_LAMBDA_FUNCTION_MEMORY_SIZE": "4096",
          "AWS_LAMBDA_RUNTIME_API": "http://[::]:9000/.rt",
          "AWS_LAMBDA_FUNCTION_NAME": "_"
        }
      },
      "args": [],
      "cwd": "${workspaceFolder}"
    }
  ]
}
```

Do you have an idea of what could be happening?
Thanks for this awesome library and the help!

**Reply by calavera** [reply]
sorry, I completely missed this comment. I'm guessing you didn't manage to figure out what the problem was, did you?

**Reply by ngryman** [reply]
No worries. No I couldn't figure out a way to make it work.

**Comment by denys-bushuliak** [comment]
I hope this trick will help you (@ngryman)

1. Start you lambda as usual `cargo lambda watch`
2. Add to VS Code configurations this (we smart folks, and know that `my-app` have to be replaced with your app name):
```json
{
  "name": "(OSX) Launch",
  "type": "lldb",
  "request": "launch",
  "program": "${workspaceRoot}/target/debug/my-app",
  "args": [],
  "cwd": "${workspaceRoot}",
  "env": {
    "AWS_LAMBDA_FUNCTION_VERSION": "1",
    "AWS_LAMBDA_FUNCTION_MEMORY_SIZE": "4096",
    "AWS_LAMBDA_RUNTIME_API": "http://[::]:9000/.rt",
    "AWS_LAMBDA_FUNCTION_NAME": "_"
  }
}
```
3. Put breakpoint in lambda function where you want.
4. Launch debugger with this configuration and open "bottle of vine".

P.S.: in log of lambda you will be overfilled with this kind of message (how to prevent it, or maybe it's normal - I don't know):
```
DEBUG pooling idle connection for ("http", [::]:9000)
DEBUG reuse idle connection for ("http", [::]:9000)
DEBUG flushed 108 bytes
DEBUG parsed 8 headers
DEBUG incoming body is empty
```

P.P.S.:
Thanks author of this article: https://www.forrestthewoods.com/blog/how-to-debug-rust-with-visual-studio-code

**Reply by westy92** [reply]
Don't forget to invoke your lambda with `cargo lambda invoke --data-ascii '{}'` or else your breakpoint won't get hit.
